---
title: "Reproducible environments in R"
subtitle: "Shifting sands to solid ground"
author: "Eric Edward Bryant"
date: "2020-01-06"
slug: "reproducible-environments-in-r"
bibliography: ../../static/references/library.bib
link-citations: yes
output:
  blogdown::html_page:
    toc: true
categories:
  - R
tags:
  - R
  - renv
  - "Reproducible research"
---


<div id="TOC">
<ul>
<li><a href="#tldr">TL;DR</a></li>
<li><a href="#why-an-r-package-version-manager-is-important">Why an R package version manager is <em>important</em></a></li>
<li><a href="#my-journey-to-renv">My journey to <span>renv</span></a></li>
<li><a href="#remaining-rough-edges">Remaining rough edges</a></li>
<li><a href="#conclusion">Conclusion</a></li>
<li><a href="#footnotes-and-references">Footnotes and references</a></li>
</ul>
</div>

<div id="tldr" class="section level1">
<h1>TL;DR</h1>
<p>The <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> package brings virtual environments to the R language.
Packages are installed into a global cache, which carefully records version
information, allowing project-specific libraries to be constructed by creating
a folder of symlinks to the appropriate package versions.
<a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> keeps package versions consistent between collaborators.
When used with versioned package repositories to enable binary package
installations, restoring a project‚Äôs package library on a new computer is a
breeze.</p>
<p>Key features:</p>
<ol style="list-style-type: decimal">
<li>Backed by <a href="https://rstudio.com">RStudio</a> as a replacement for <a href="https://rstudio.github.io/packrat">packrat</a>, so it should receive
continued development and support</li>
<li>Global package cache reduces wasted space‚Äîa big plus when using large
packages (e.g.¬†genome packages from Bioconductor)</li>
<li>Plays well with Python virtual environments (see¬†<code>renv::use_python</code>)</li>
</ol>
<p>Try it out on an existing project:</p>
<pre class="r"><code>install.packages(&quot;renv&quot;)
renv::init()</code></pre>
</div>
<div id="why-an-r-package-version-manager-is-important" class="section level1">
<h1>Why an R package version manager is <em>important</em></h1>
<p>In short, R package updates frequently introduce breaking changes.
Without careful R package management, analyses will not be easily reproducible
in the future.
This does not mean we should avoid using packages; package version management
is a solvable problem.</p>
<p>I was introduced to the <a href="https://r-project.org">R</a> programming language in the spring of 2012 as part
of a class on molecular systems biology during the first year of my PhD.
The experience was a bit rough at first, but I was determined to learn a
language that could help me with my summer rotation project.
And, since it was required for my coursework, it was only natural to focus on
learning R.</p>
<p>Using R, I encountered a number of pain points, and often considered jumping
ship to focus on learning other popular languages.
But, every time I was about to turn away from R, somebody would solve my
problem in a new package, softening R‚Äôs rough edges.
All I needed to do was <code>install.packages(&quot;solve.erics.problem.now&quot;)</code>, and I was
happy as a clam<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<p>Some notable examples of this:</p>
<ul>
<li>I need to track a project‚Äôs progress with dynamic reports‚Äî<a href="https://yihui.org/knitr">knitr</a>, (2012)</li>
<li>I need these reports to have proper bibliographies‚Äî<a href="https://rmarkdown.rstudio.com">rmarkdown</a>, (2014)
an integration of <a href="https://yihui.org/knitr">knitr</a> and <a href="https://pandoc.org">pandoc</a>, (2006)</li>
<li>I need better cross-referencing for my thesis‚Äî<a href="https://bookdown.org/yihui/bookdown">bookdown</a>, (2016)</li>
<li>I need to do image processing‚Äî<a href="https://bioconductor.org/packages/release/bioc/vignettes/EBImage/inst/doc/EBImage-introduction.html">EBImage</a>, (2010)</li>
<li>I need to quickly understand analysis code I wrote 6 months
ago‚Äî<a href="https://dplyr.tidyverse.org">dplyr</a>, (2014)</li>
<li>I need to iterate as quickly as I can think for data
visualization‚Äî<a href="https://ggplot2.tidyverse.org">ggplot2</a>, (2007)</li>
<li>I need some way for my students to enter data, and to check and version
control that data for analyses‚Äî<a href="https://github.com/jennybc/googlesheets">googlesheets</a>, (2015)</li>
<li>I have a bad bottleneck in my code‚Äî<a href="http://dirk.eddelbuettel.com/code/rcpp.html">RCpp</a>, (2008)</li>
<li>I want to do some network analysis‚Äî<a href="https://igraph.org/r/">igraph</a>, (2006)</li>
<li>I need more perplexity in my life‚Äî<a href="https://github.com/jkrijthe/Rtsne">Rtsne</a>, (2014)</li>
</ul>
<p>This list could go on forever.
There are seemingly innumerable problems that have been solved by an R package.
Installing and using packages in R is dead simple.
At this moment I have 230 packages installed on my system.
And that is me being fairly conservative when adopting new packages!</p>
<p>Unfortunately, casually installing and using packages comes at a cost
(see <a href="https://www.tidyverse.org/blog/2019/05/itdepends">here</a> for some
insightful discussion).
If you are not careful, you may inadvertently step into <a href="https://en.wikipedia.org/wiki/Dependency_hell">dependency hell</a>,
which is often not obvious until you need to share your project with a
collaborator.
They will ask,</p>
<blockquote>
<p>What do I need to install?</p>
</blockquote>
<p>You respond,</p>
<blockquote>
<p>These 130 packages are required.
90 can be installed as usual.
Wait, actually, I have some bioconductor packages in there, so I guess you
should use their package installer?
Oh and I‚Äôm using the development version of this and that package, so be sure
to install from GitHub for those.
And, I seem to remember there were some breaking changes to such and such
package, so you will need to install the old version until I get around to
fixing that.
Oh and I used a new function recently added to R, so be sure to update
R‚Äîyou might need to re-install everything if there has been a minor
version change.
Don‚Äôt worry, I wrote an install script.
You ran into an error?
You‚Äôre on Windows at work, macOS at home and Ubuntu at the gym?
Well, then it‚Äôs probably a problem with your setup.
I recommend testing installation at work, home, and gym.
Do you have Rtools/XCode/build-essential installed?
Ooooh, actually, I forgot that I started using <code>superAwesomeNewPackage</code>.
Be sure to install that as well, I‚Äôll update the install script.
Simply run that script and everything should work great.&quot;</p>
</blockquote>
<p>And they never want to collaborate with you again.
Of course, this is an exaggeration.
Nevertheless, complicated installation instructions and package updates that
introduce breaking changes can make reproducing an old analysis feel nearly
impossible, especially if package versions were not meticulously recorded.
Certainly, for an infrequent R user, old analyses that rely on many packages
are often not reproducible<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<p>So, this is the most pressing pain point I have with R right now:</p>
<ul>
<li>I need to stabilize package versions to protect my projects from <a href="https://en.wikipedia.org/wiki/Software_rot">bit rot</a>.
I also need to make it easy for a collaborator to install a specific
constellation of fixed package versions in a way that does not break their
existing setup, or cause too much administrative burden.
In short, <strong>I want to move my projects off of shifting sands and onto solid
ground.</strong></li>
</ul>
</div>
<div id="my-journey-to-renv" class="section level1">
<h1>My journey to <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a></h1>
<p>Over time I have made incremental progress towards solving the package
versioning problem for myself.
The first part of the solution was to begin using (semi-)stable
repositories<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>.
I did this by adding the following to my <code>~/.Rprofile</code> (an R script that gets
sourced when an R session starts).</p>
<pre class="r"><code>options(
  repos = c(
    CRAN          = &quot;https://cran.microsoft.com/snapshot/2019-12-12&quot;,
    BioCsoft      = &quot;https://bioconductor.org/packages/3.10/bioc&quot;,
    BioCann       = &quot;https://bioconductor.org/packages/3.10/data/annotation&quot;,
    BioCexp       = &quot;https://bioconductor.org/packages/3.10/data/experiment&quot;,
    BioCworkflows = &quot;https://bioconductor.org/packages/3.10/workflows&quot;
  )
)</code></pre>
<p>Once this is set, <code>install.packages()</code> will install the package versions from
CRAN that existed on <code>2019-12-12</code>, and from bioconductor version <code>3.10</code>.
Of course, the appropriate date and bioconductor version depends on the version
of R currently in use.
So, my <code>~/.Rprofile</code> dynamically sets the appropriate repositories depending
on the active version of R, which looks roughly like the following.</p>
<pre class="r"><code>.First &lt;- function() .repos_set_stable()

# ---- Hidden convenience functions for personal use ----
.repos_set_stable &lt;- function(r_version = getRversion()) {
  options(
    repos = switch(
      as.character(r_version),
      # Remember to update 3.6.2 to R-4.0.0 release date
      &quot;3.6.2&quot; = .repos_stable(&quot;2019-12-12&quot;, bioc = &quot;3.10&quot;),
      &quot;3.5.3&quot; = .repos_stable(&quot;2019-12-12&quot;, bioc = &quot;3.8&quot;),
      &quot;3.4.4&quot; = .repos_stable(&quot;2019-03-11&quot;, bioc = &quot;3.6&quot;),
      &quot;3.3.3&quot; = .repos_stable(&quot;2018-03-19&quot;, bioc = &quot;3.4&quot;)
    )
  )
}

.repos_stable &lt;- function(date, bioc,
                          url_mran = &quot;https://cran.microsoft.com/snapshot&quot;,
                          url_bioc = &quot;https://bioconductor.org/packages&quot;) {
  c(CRAN          = file.path(url_mran, date, fsep = &quot;/&quot;),
    BioCsoft      = file.path(url_bioc, bioc, &quot;bioc&quot;, fsep = &quot;/&quot;),
    BioCann       = file.path(url_bioc, bioc, &quot;data/annotation&quot;, fsep = &quot;/&quot;),
    BioCexp       = file.path(url_bioc, bioc, &quot;data/experiment&quot;, fsep = &quot;/&quot;),
    BioCworkflows = file.path(url_bioc, bioc, &quot;workflows&quot;, fsep = &quot;/&quot;)
  )
}</code></pre>
<p>For personal use, this setup goes a long way towards saving my sanity.
Any given version of R will have one officially supported constellation
of package versions from its set of supported stable repositories, and most of
these versions will be available as fast installing pre-compiled binaries on
macOS and Windows.
If I want to run an old analysis of mine, I switch to an appropriate version of
R recorded in the project, and everything should work.
If I am missing a package, I can use <code>install.packages()</code> as per usual without
too much fuss.</p>
<p>The problem comes when imposing my workflow onto collaborators.
Some collaborators, understandably, want the latest and greatest of
everything, and certainly do not want to overwrite any previously installed
packages with out-of-date versions.
And, if I want to reproduce other people‚Äôs work, then I end up having to
install versions they used, which ends up polluting my package library with
‚Äúunsupported‚Äù (by me) versions.
My solution to this was to isolate my ‚Äústable‚Äù packages into their own
library<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>.
I cobbled together my solution into a package that I called <a href="https://github.com/ericedwardbryant/rprofile">rprofile</a> to
streamline the creation and configuration of <code>.Rprofile</code>s that would stabilize
project package versions (and versioned libraries could be shared across
projects).
This solution was OK, but not great.
I was happy with it, and my collaborators seemed to tolerate it, even though
it required a small setup cost and yelled at them when they installed packages
into the default system library (nobody likes being yelled at).</p>
<p>Recently, however, I was helping a collaborator get set up with a project on a
Windows computer and ran into some issues<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<p>The <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> package arrived on CRAN towards the end of October 2019, just in
time to solve my problem (mostly).
I still find it useful to use stable repositories, but now, when I am starting
a project, I run <code>renv::init()</code>, and <code>renv::snapshot()</code> and my project gets
set up with isolated package libraries.
Best of all, a nice record of my repositories and package versions used in the
project gets recorded in a JSON formatted file called <code>renv.lock</code>.
It looks something like the following.</p>
<pre class="json"><code>{
  &quot;R&quot;: {
    &quot;Version&quot;: &quot;3.6.2&quot;,
    &quot;Repositories&quot;: [
      {
        &quot;Name&quot;: &quot;CRAN&quot;,
        &quot;URL&quot;: &quot;https://cran.microsoft.com/snapshot/2019-12-12&quot;
      },
      {
        &quot;Name&quot;: &quot;BioCsoft&quot;,
        &quot;URL&quot;: &quot;https://bioconductor.org/packages/3.10/bioc&quot;
      },
      {
        &quot;Name&quot;: &quot;BioCann&quot;,
        &quot;URL&quot;: &quot;https://bioconductor.org/packages/3.10/data/annotation&quot;
      },
      {
        &quot;Name&quot;: &quot;BioCexp&quot;,
        &quot;URL&quot;: &quot;https://bioconductor.org/packages/3.10/data/experiment&quot;
      },
      {
        &quot;Name&quot;: &quot;BioCworkflows&quot;,
        &quot;URL&quot;: &quot;https://bioconductor.org/packages/3.10/workflows&quot;
      }
    ]
  },
  &quot;Packages&quot;: {
    &quot;BH&quot;: {
      &quot;Package&quot;: &quot;BH&quot;,
      &quot;Version&quot;: &quot;1.69.0-1&quot;,
      &quot;Source&quot;: &quot;Repository&quot;,
      &quot;Repository&quot;: &quot;CRAN&quot;,
      &quot;Hash&quot;: &quot;f4605d46264b35f53072fc9ee7ace15f&quot;
    }
  }
...
}</code></pre>
<p>Getting this project working on another computer just requires calling
<code>renv::restore()</code>, and, if the R version doesn‚Äôt match the version recorded
here, a warning is issued on startup.
And, the package library will be fully isolated.
For example, the libraries used as I write this post are shown below.
Note the system default library that contains all the base R packages is not
present (instead that second library just has symlinks to base R packages).
No more worrying about polluting the system library!</p>
<pre class="r"><code>.libPaths()</code></pre>
<pre><code>## [1] &quot;/Users/ericbryant/Dropbox/GitHub/ericbryantphd.com/renv/library/R-3.6/x86_64-apple-darwin15.6.0&quot;
## [2] &quot;/private/var/folders/8j/_6xf2s2d2c75d83_y37cq0100000gn/T/RtmpHavfa9/renv-system-library&quot;</code></pre>
<p>Also, because I use stable repositories, and because <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> records these in
<code>renv.lock</code>, binary packages are available for old package versions without my
collaborators having to do <em>any</em> configuration.</p>
<p>Perhaps the nicest feature of <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> is that it installs packages into a global
cache, so that multiple projects that use the same version of a package can
share these resources.
In fact, the project‚Äôs private package library, stored in <code>renv/library</code>, just
contains symlinks to the global cache, so there is almost no overhead to use
<a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> on every new project.
All in, <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> requires very little investment to use and is certain to save
me from time breaking all my things.</p>
</div>
<div id="remaining-rough-edges" class="section level1">
<h1>Remaining rough edges</h1>
<p><a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> solves R package stabilization nicely, but, of course, there are still a
few rough edges (some of these are clearly out-of-scope for <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> to solve).
I‚Äôm recording these here mostly as a reminder to myself of some avoidable
pitfalls when using <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> to create reproducible analyses.</p>
<ol style="list-style-type: decimal">
<li><p>CRAN does not maintain binaries for old packages, so, when running
<code>renv::restore()</code>, if a collaborator must install an old package, they will
often need to install <a href="https://support.rstudio.com/hc/en-us/articles/200486498-Package-Development-Prerequisites">package development prerequisites</a> first to be able
to compile source packages.
This is not a major issue for most packages, but can lead to installation
failures for packages that require esoteric system dependencies.
This is why I still use stable repositories that host binaries for the
package versions I use.
The downside of this is that I have to be diligent about managing stable
repositories for every R upgrade
(not a huge deal, just an update to my <code>~/.Rprofile</code> described above).
The other downside is that I rely on Microsoft and Bioconductor to host
these repositories forever, which is kind of a big ask.
ü§∑</p></li>
<li><p>When searching a project, depending on the text editor in use, the entire
contents of <code>renv/library</code> will also be searched, which slows down search
and clutters results with unimportant hits from the package library.
Usually this just requires being a little more specific about where to
perform a search, so not a huge deal.</p></li>
<li><p>Many packages provide wrappers for APIs that are inherently out of their
control and therefore inherently unstable.
<a href="https://github.com/jennybc/googlesheets">googlesheets</a> is a good example of this problem.
This package provides access to Google Sheets, which wraps version 3 of the
Sheets API.
Well, now there is a <a href="https://googlesheets4.tidyverse.org">googlesheets4</a>, which wraps version 4 and version 3
will stop working in March 2020.
Thus, using <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> to track and install old versions of <a href="https://github.com/jennybc/googlesheets">googlesheets</a> is
kinda pointless and potentially harmful, since the old version will no
longer work.
For this reason, it is probably best not to strictly rely on packages that
wrap APIs.
I will still use <a href="https://googlesheets4.tidyverse.org">googlesheets4</a> because it is extremely useful when you
need to have a simple data entry method and want to retrieve that data
easily and programmatically, but I end up committing the imported data to
my project such that I can always re-run my analysis without strictly
requiring <a href="https://googlesheets4.tidyverse.org">googlesheets4</a> and a stable API from Google.
Fortunately, <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> does provide a mechanism to exclude packages from
<code>renv.lock</code>.
I like to add these packages to <code>renv/settings.dcf</code> with a line like the
following:</p>
<pre><code>ignored.packages: googlesheets</code></pre>
<p>This prevents <code>renv::snapshot()</code> from adding <a href="https://github.com/jennybc/googlesheets">googlesheets</a> to <code>renv.lock</code>.
To be honest, I don‚Äôt even really worry too much about this because I don‚Äôt
consider it a huge problem to have a broken API wrapping package in my
project library, so long as it isn‚Äôt strictly necessary to use that package
to clean and rebuild the project.</p></li>
<li><p>I have run into some problems using <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> with <a href="https://dplyr.tidyverse.org">devtools</a>, but I think
this is going to be fixed soon, so I am not going to elaborate on this too
much (see <a href="https://github.com/rstudio/renv/issues/297">here</a> for details).</p></li>
<li><p>Rough edge #4 is also related to a small issue I have had with large
projects that contain <a href="https://rmarkdown.rstudio.com">rmarkdown</a> files in subdirectories.
The crux of this issue is that <a href="https://rmarkdown.rstudio.com">rmarkdown</a>/<a href="https://yihui.org/knitr">knitr</a> use an independent
R process that is launched from the directory containing the <a href="https://rmarkdown.rstudio.com">rmarkdown</a>
file.
When this happens, the project‚Äôs <code>.Rprofile</code> is not loaded, which means
<a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> is not initiated.
There are a couple ways to solve this problem.
Currently, I add the following to the first code chunk of my reports:</p>
<pre><code>knitr::opts_knit$set(root.dir = here::here())</code></pre>
<p>This tells <a href="https://yihui.org/knitr">knitr</a> to run code using an R process launched from the project
level directory, which is fund by <code>here::here()</code>
(which happens to sound pleasingly congratulatory).
Let me know if you have a good way to set this globally for all <a href="https://rmarkdown.rstudio.com">rmarkdown</a>
files in a project!</p></li>
<li><p><a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> does not perfectly solve the ‚Äúreproducible environment‚Äù problem.
For example, <a href="https://rmarkdown.rstudio.com">rmarkdown</a> has a system requirement for the <a href="https://pandoc.org">pandoc</a> command
line utility.
<a href="https://rmarkdown.rstudio.com">rmarkdown</a> searches for an existing system installation of <a href="https://pandoc.org">pandoc</a>, and
for a <a href="https://pandoc.org">pandoc</a> that comes bundled with <a href="https://rstudio.com">RStudio</a>.
<a href="https://rmarkdown.rstudio.com">rmarkdown</a> then chooses the latest version of <a href="https://pandoc.org">pandoc</a> that exists between
the system and <a href="https://rstudio.com">RStudio</a> installations.
This caused a problem when <a href="https://pandoc.org">pandoc</a> made major breaking changes with the
release of version 2 in 2017.
Now, all of my old <a href="https://rmarkdown.rstudio.com">rmarkdown</a> reports require me to do a back-flip to
re-compile.
This was especially problematic for me because I upgraded RStudio while a
paper was in review only to find that my manuscript would no longer build
because of the breaking changes in <a href="https://pandoc.org">pandoc</a>.
After a brief panic attack, I realized that I should be freezing <a href="https://pandoc.org">pandoc</a>
the same way <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> freezes R packages.
Unfortunately, since <a href="https://rmarkdown.rstudio.com">rmarkdown</a> always picks the latest version it finds
from the system or RStudio, I have to use the following magic words in each
project‚Äôs <code>.Rprofile</code> to override this behavior:</p>
<pre><code># Add ./renv/bin to PATH, and override RStudio pandoc
local({
  bin &lt;- normalizePath(&quot;renv/bin&quot;)
  Sys.setenv(
    PATH = paste(bin, Sys.getenv(&quot;PATH&quot;), sep = &quot;:&quot;),
    RSTUDIO_PANDOC = bin
  )
})</code></pre>
<p>This basically adds <code>renv/bin</code> to the front of the <code>PATH</code> variable, such
that programs installed here get priority, and resets the
<code>RSTUDIO_PANDOC</code> environment variable to <code>renv/bin</code> such that <a href="https://rmarkdown.rstudio.com">rmarkdown</a>
will only find <a href="https://pandoc.org">pandoc</a> here.
I currently copy all system requirements to <code>renv/bin</code> when a project has
matured and is ready to be archived, but even then, I often have to figure
out how to get R packages to look here for their system requirements.
For example, to get <a href="https://bookdown.org/yihui/blogdown">blogdown</a> to use <a href="https://gohugo.io">hugo</a> in <code>renv/bin</code> I had to use:</p>
<pre><code>options(blogdown.hugo.dir = normalizePath(&quot;renv/bin&quot;))</code></pre>
<p>This <code>renv/bin</code> solution of mine feels like an ad-hoc error-prone field
repair that is almost certainly not platform independent and doesn‚Äôt make
it easy to install these requirements on a new system.</p></li>
</ol>
</div>
<div id="conclusion" class="section level1">
<h1>Conclusion</h1>
<p>In conclusion, <a href="https://rstudio.github.io/renv/articles/renv.html">renv</a> is my favorite new development in the R world and I
intend to begin using it with all of my projects including some of my archived
work.
Rough edge #6 is my remaining big hurdle towards full reproducibility.
Eventually I will leap this hurdle with <a href="https://www.docker.com">docker</a>, as that seems to be the way
the world is moving, but right now, asking my collaborators to use <a href="https://www.docker.com">docker</a>
feels like asking them to travel to my project which is currently hosted on
the moon.
I hope to change that feeling soon!</p>
<p>If you have any questions or comments, please reach out to me on <a href="https://twitter.com/ericedwrdbryant">twitter</a>!</p>
</div>
<div id="footnotes-and-references" class="section level1 unnumbered">
<h1>Footnotes and references</h1>
<div id="refs" class="references">
<div id="ref-2017-09-Billon-Bryant-Ciccia">
<p>Billon, Pierre, Eric E Bryant, Sarah A Joseph, Tarun S Nambiar, Samuel B Hayward, Rodney Rothstein, and Alberto Ciccia. 2017. ‚ÄúCRISPR-Mediated Base Editing Enables Efficient Disruption of Eukaryotic Genes Through Induction of STOP Codons.‚Äù <em>Molecular Cell</em> 67 (6): 1068‚Äì1079.e4. <a href="https://doi.org/10.1016/j.molcel.2017.08.008">https://doi.org/10.1016/j.molcel.2017.08.008</a>.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>I am not really sure how happy clams are, but I think probably this
happy ü§©<a href="#fnref1" class="footnote-back">‚Ü©</a></p></li>
<li id="fn2"><p>My condolences for anyone attempting to reproduce my own analysis, with 90
package dependencies, described in <span class="citation">Billon et al. (<a href="#ref-2017-09-Billon-Bryant-Ciccia">2017</a>)</span>, available
<a href="https://github.com/CicciaLab/iSTOP-paper">here</a><a href="#fnref2" class="footnote-back">‚Ü©</a></p></li>
<li id="fn3"><p>Repositories are where R packages are downloaded from<a href="#fnref3" class="footnote-back">‚Ü©</a></p></li>
<li id="fn4"><p>A library is a directory where R packages are installed.
R packages themselves are directories containing code and documentation.<a href="#fnref4" class="footnote-back">‚Ü©</a></p></li>
<li id="fn5"><p>Ultimately my <a href="https://github.com/ericedwardbryant/rprofile">rprofile</a> issue was just that Windows does not allow
writing to the directory that R lives in, which is where <a href="https://github.com/ericedwardbryant/rprofile">rprofile</a> tried to
make the versioned libraries by default.
It was an easy fix, but le sigh üò≠<a href="#fnref5" class="footnote-back">‚Ü©</a></p></li>
</ol>
</div>
